File: app/python/plugins/storage.py
from lattice import Capability, ModuleSpec


def start_storage(ctx):
    class StorageV1:
        def __init__(self):
            self._data = {}

        def put(self, key, value):
            print(f"[storage-v1] put {key}={value}")
            self._data[key] = value

        def get(self, key):
            print(f"[storage-v1] get {key}")
            return self._data.get(key)

    ctx.bind(Capability("storage", (1, 0), frozenset({"memory"})), StorageV1())


REGISTER_MODULES = [
    ModuleSpec(
        name="storage-v1",
        provides=[Capability("storage", (1, 0), frozenset({"memory"}))],
        requires=[],
        start=start_storage,
    )
]

REGISTER_ADAPTERS = []

File: typescript/src/plugins/storage.ts
import type { ModuleSpec } from "../core.js";

export const REGISTER_MODULES: ModuleSpec[] = [
  {
    name: "storage-v1",
    provides: [{ name: "storage", version: [1, 0], traits: ["memory"] }],
    requires: [],
    start(ctx) {
      const data = new Map<string, unknown>();
      ctx.bind({ name: "storage", version: [1, 0], traits: ["memory"] }, {
        put(key: string, value: unknown) {
          console.log(`[storage-v1] put ${key}=${JSON.stringify(value)}`);
          data.set(key, value);
        },
        get(key: string) {
          console.log(`[storage-v1] get ${key}`);
          return data.get(key);
        },
      });
    },
  },
];

export const REGISTER_ADAPTERS = [];
